### PR TITLE
Iow 652 date acc cd cannot be null

### PR DESCRIPTION
**Pull request by ee-usgs**
Before making a pull request
----------------------------

- [x] Make sure all tests run
- [x] Update the changelog appropriately

Title
-----------
Brief description of changes. Reference the JIRA ticket if appropriate

Description
-----------
If no JIRA ticket is referenced, describe the changes made. Note anything that you want the reviewers to know while
reviewing your pull request

After making a pull request
---------------------------
- [x] If appropriate, put the link to the PR in the JIRA ticket
- [x] Assign someone to review unless the change is trivial

### CODE REVIEW

File: src/main/java/gov/usgs/wma/waterdata/DateTimeAccuracy.java [hunk 1]
```diff
@@ -77,4 +77,14 @@ public static DateTimeAccuracy parse(final String sourceText) {
...
+
+	/**
+	 * This is the string we expect to find in the field_visit_comment to indicate this
+	 * DateTimeAccuracy
+	 * @return
+	 */
+	public String getMatchString() {
+		return matchString;
+	}
+
```
Comment (ee-usgs) [inline src/main/java/gov/usgs/wma/waterdata/DateTimeAccuracy.java:89]: This is really just for the unit tests to be able to access this string

File: src/main/java/gov/usgs/wma/waterdata/DiscreteGroundWaterRowMapper.java [hunk 2]
```diff
@@ -10,14 +10,14 @@
...
 
-		DateTimeAccuracy dta = DateTimeAccuracy.parse(rs.getString("field_visit_comments"));
+		DiscreteGroundWaterRules rules = new DiscreteGroundWaterRules();
 
 		DiscreteGroundWater discreteGroundWater = new DiscreteGroundWater();
 		discreteGroundWater.setFieldVisitIdentifier(rs.getString("field_visit_identifier"));
 		discreteGroundWater.setLocationIdentifier(rs.getString("location_identifier"));
 		discreteGroundWater.setAgencyCode(rs.getString("agency_code"));
-		discreteGroundWater.setDateTimeAccuracyCode(dta.getCode());
-		discreteGroundWater.setDateTimeAccuracyText(dta.getText());
```
Comment (ee-usgs) [inline src/main/java/gov/usgs/wma/waterdata/DiscreteGroundWaterRowMapper.java:20]: Previously, there was some business logic in here, parsing the comment field and assigning the dta code and text based on it.  Now that has all moved to the 'rules' class and these two derived fields are left as null until the rules are applied.

File: src/main/java/gov/usgs/wma/waterdata/DiscreteGroundWaterRowMapper.java [hunk 3]
```diff
@@ -47,6 +47,9 @@ public DiscreteGroundWater mapRow(ResultSet rs, int rowNum) throws SQLException
 		discreteGroundWater.setGroundWaterMeasurement(rs.getString("ground_water_measurement"));
 		discreteGroundWater.setDatum(rs.getString("datum"));
 
+		//Applies all business rules
+		rules.apply(discreteGroundWater);
+
```
Comment (ee-usgs) [inline src/main/java/gov/usgs/wma/waterdata/DiscreteGroundWaterRowMapper.java:52]: Is there a better place to apply business rules??

File: src/test/java/gov/usgs/wma/waterdata/BaseTestDao.java [hunk 4]
```diff
@@ -55,6 +55,8 @@ public void setup() {
 		discreteGroundWater1.setFieldVisitIdentifier("0c8d1725-db51-47d9-aabb-a1dd9d1bdf08");
 		discreteGroundWater1.setLocationIdentifier(LOCATION_IDENTIFIER_1);
 		discreteGroundWater1.setAgencyCode("USGS");
+		discreteGroundWater1.setDateTimeAccuracyCode(DateTimeAccuracy.MINUTE.getCode());
+		discreteGroundWater1.setDateTimeAccuracyText(DateTimeAccuracy.MINUTE.getText());
```
Comment (ee-usgs) [inline src/test/java/gov/usgs/wma/waterdata/BaseTestDao.java:59]: The sample dom object built up in this base test class are supposed to be after they have been loaded by the RowMapper, so they need to already be in a state where they follow the rules.

File: src/test/java/gov/usgs/wma/waterdata/DiscreteGroundWaterRulesTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	@Test
+	void nullFieldVisitCommentsShouldParseToMINUTEDateTimeAccuracyCode() {
+
+		dga.setFieldVisitComments(null);
+
+		rules.apply(dga);
+		assertEquals(DateTimeAccuracy.MINUTE.getCode(), dga.getDateTimeAccuracyCode());
+		assertEquals(DateTimeAccuracy.MINUTE.getText(), dga.getDateTimeAccuracyText());
+	}
+}
```
Comment (ee-usgs) [inline src/test/java/gov/usgs/wma/waterdata/DiscreteGroundWaterRulesTest.java:58]: Another nice part about business rules being in 'Java land' is that I can really unit test them completely - way less hassle then requiring db level tests for business rules.

File: src/main/java/gov/usgs/wma/waterdata/DiscreteGroundWaterRules.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+				domObj.setDateTimeAccuracyText(dta.getText());
+			} else {
+
+				dta = DateTimeAccuracy.MINUTE;
+				domObj.setDateTimeAccuracyCode(dta.getCode());
+				domObj.setDateTimeAccuracyText(dta.getText());
+			}
+		}
+	}
+}
```
Comment (ee-usgs) [inline src/main/java/gov/usgs/wma/waterdata/DiscreteGroundWaterRules.java:34]: If we follow this model, we can have all of our business rules in one place with lots of room for grouping and documenting.

**Review by ssoper-usgs — APPROVED** [review]

**Review by kkehl-usgs — APPROVED** [review]

**Review by orosz-usgs — APPROVED** [review]
Looks good.

**Pull request merged**